### PR TITLE
Added logical byte functions and in-byte sequence

### PR DESCRIPTION
**Pull request by dys-bigwig**

### CODE REVIEW

File: private/byte.rkt [hunk 1]
```diff
@@ -8,7 +8,14 @@
...
   [byte-drop-rightmost-bits (-> byte? (integer-in 0 8) byte?)]
-  [byte-ref (-> byte? (integer-in 0 7) bit?)]))
+  [byte-ref (-> byte? (integer-in 0 7) bit?)]
+  [byte-and (-> byte? byte? byte?)]
+  [byte-or (-> byte? byte? byte?)]
+  [byte-not (-> byte? byte?)]
+  [byte-xor (-> byte? byte? byte?)]
+  [byte-nand (-> byte? byte? byte?)]
+  [byte-nor (-> byte? byte? byte?)]
+  [in-byte (-> byte? sequence?)]))
```
Comment (jackfirth) [inline private/byte.rkt]: For `in-byte`, could you strengthen the contract from `sequence?` to `(sequence/c bit?)`?

File: private/byte.rkt [hunk 2]
```diff
@@ -109,3 +116,279 @@
...
+
+(define (byte-not b)
+  (apply byte
+         (for/list ([i (in-range 8)])
+           (bitwise-xor (byte-ref b i) 1)))) 
+
+(module+ test
+  (test-case
+    "byte-not"
+    (for ([x (in-range 256)])
```
Comment (jackfirth) [inline private/byte.rkt]: Another test we could add for `byte-not`: a byte and its negation should always add up to 255.

```
(for ([x (in-range 256)])
  (check-equal? (+ x (byte-not x)) 255))
```

File: private/byte.rkt [hunk 3]
```diff
@@ -109,3 +116,279 @@
...
+
+(define (byte-xor left right)
+  (apply byte
+         (for/list ([i (in-range 8)])
+           (bitwise-xor (byte-ref left i)
+                        (byte-ref right i))))) 
+(module+ test
+  (test-case
+    "byte-xor"
+    (for ([x (in-range 256)])
```
Comment (jackfirth) [inline private/byte.rkt]: Another test: a byte XOR'd with itself always produces zero.

```
(for ([x (in-range 256)])
  (check-equal? (byte-xor x x) 0))
```

File: private/byte.rkt [hunk 4]
```diff
@@ -109,3 +116,279 @@
...
+                    (byte-ref (byte-not x) 5))
+
+      (check-equal? (byte-ref (byte-xor x 0) 6)
+                    (byte-ref x 6))
+      (check-equal? (byte-ref (byte-xor x 255) 6)
+                    (byte-ref (byte-not x) 6))
+
+      (check-equal? (byte-ref (byte-xor x 0) 7)
+                    (byte-ref x 7))
+      (check-equal? (byte-ref (byte-xor x 255) 7)
```
Comment (jackfirth) [inline private/byte.rkt]: I think many of the xor tests can be combined into a few simpler tests, like this:

```
(for ([x (in-range 256)])
  (check-equal? (byte-xor x 0) x)
  (check-equal? (byte-xor x 255) (byte-not x)))
```

That seems easier than doing eight separate checks (one for each bit) for both of these properties we want to test.

File: private/byte.rkt [hunk 5]
```diff
@@ -109,3 +116,279 @@
...
+         (for/list ([i (in-range 8)])
+           (bitwise-and (byte-ref left i)
+                        (byte-ref right i))))) 
+
+(module+ test
+  (test-case
+    "byte-and"
+    (for ([x (in-range 256)])
+      (check-equal? (byte-ref (byte-and x (byte 0 1 1 1 1 1 1 1)) 0) 0)
+      (check-equal? (byte-ref (byte-and (byte 0 1 1 1 1 1 1 1) x) 0) 0)
```
Comment (jackfirth) [inline private/byte.rkt]: Instead of having each check twice here, we could have a separate loop that checks the operation is always commutative.

```
(for* ([x (in-range 256)] [y (in-range 256)])
  (check-equal? (byte-and x y) (byte-and y x)))
```

You'll need to use `for*` instead of `for`. The `for` macro steps over each loop variable together, whereas `for*` acts like a nested for. The above `for*` loop is equivalent to this:

```
(for ([x (in-range 256)])
  (for ([y (in-range 256)])
    (check-equal? (byte-and x y) (byte-and y x))))
```

File: private/byte.rkt [hunk 6]
```diff
@@ -109,3 +116,279 @@
...
+      (check-equal? (byte-ref (byte-xnor x 255) 6)
+                    (byte-ref x 6))
+
+      (check-equal? (byte-ref (byte-xnor x 0) 7)
+                    (byte-ref (byte-not x) 7))
+      (check-equal? (byte-ref (byte-xnor x 255) 7)
+                    (byte-ref x 7)))))
+                    
+(define (in-byte b)
+  (make-do-sequence
```
Comment (jackfirth) [inline private/byte.rkt]: The `make-do-sequence` function is super complicated and hard to use. If you like, instead you could return the list directly because lists implement the sequence interface.

```
(define (in-byte b)
  (list (byte-ref b 0) (byte-ref b 1) ...))
```

File: private/byte.rkt [hunk 7]
```diff
@@ -109,3 +116,279 @@
...
+             #false
+             #false))))
+
+(module+ test
+  (test-case
+    "in-byte sequence"
+    (for ([x (in-range 256)])
+      (for ([b (in-byte x)]
+            [i (in-range 8)])
+        (check-equal? b (byte-ref x i))))))
```
Comment (jackfirth) [inline private/byte.rkt]: Nicely done on this test! Great work.

**Comment by jackfirth** [comment]
I looked a bit more into why the tests aren't running for you, and I think it's because pull requests from forked repositories only get CI if _both_ the owners of the original repo and the forked repo are enrolled in the GitHub Actions beta. So if you'd like to see the tests run, there's a few options:

- You can [sign up for the beta][gh-actions].
- You can wait until November 13th, which is when GitHub Actions becomes generally available.
- I can add you as a contributor to this repository and you can work on a branch of this repo instead of on your own fork. Pull requests within this repo will trigger actions, since only the owner of the repo needs to be in the beta, not every single contributor.

Let me know if you have a preference. November isn't that far away so just doing nothing and not worrying about it for now is a fine choice.

[gh-actions]: https://github.com/features/actions

File: private/byte.rkt (hunk 1 above)
Comment (dys-bigwig) [reply private/byte.rkt]: I've strengthened the contract to (sequence/c bit?), and also added a #:min-count argument with a value of 8, being that bytes are always 8-bits (and thus values) long. Let me know if you don't think the min-count arg is suitable.
Thanks.

Comment (jackfirth) [reply private/byte.rkt]: That's perfect, thanks!

File: private/byte.rkt [hunk 8]
```diff
@@ -109,3 +117,147 @@
...
+(define (byte-or left right)
+  (apply byte
+         (for/list ([i (in-range 8)])
+           (bitwise-ior (byte-ref left i)
+                       (byte-ref right i))))) 
+
+(module+ test
+  (test-case
+    "byte-or"
+    (for ([x (in-range 256)] [y (in-range 256)])
```
Comment (jackfirth) [inline private/byte.rkt]: I think the `y` variable in this loop is unused.

File: private/byte.scrbl [hunk 9]
```diff
@@ -5,6 +5,7 @@
                      rebellion/binary/bit
                      rebellion/binary/byte)
           (submod rebellion/private/scribble-evaluator-factory doc)
+          (only-in racket/base for/vector)
```
Comment (jackfirth) [inline private/byte.scrbl]: I think this import can be removed.

File: private/byte.scrbl [hunk 10]
```diff
@@ -70,3 +71,96 @@ integer between @racket[0] and @racket[255].
...
+ @(examples
+   #:eval (make-evaluator) #:once
+   (code:comment "set the least-significant bit")
+   (byte-or (byte 1 0 0 0 0 0 0 0) (byte 0 0 0 0 0 0 0 1))
+   (code:comment "set bits 5, 2, and 1")
+   (byte-or (byte 0 0 1 0 1 0 1 0) (byte 0 0 1 0 0 0 1 1)))
+
+@defproc[(byte-not [b byte?])
+         byte?]{
+ Performs a bitwise or operation on @racket[b], inverting all bits.
```
Comment (jackfirth) [inline private/byte.scrbl]: s/or/not/

**Comment by jackfirth** [comment]
@dys-bigwig Just a heads up: I'm planning to merge this later tonight (it's 1:43 PM in my timezone right now).
